tests/virt_autotest/setup_dns_service.pm

```diff
@@ -0,0 +1,223 @@
+              3600              ; retry
+              3600000           ; expire
+              86400 )           ; minimum
+
+
+\; ----------- ENREGISTREMENTS -----------
+\@                       IN NS                   ns.$dns_domain_forward.
+\;\;
+\$(echo -e $dns_records_reverse)
+\;\;
+\; ----------- ENREGISTREMENTS SPECIAUX -----------";
+    my $set_dns_reverse_zone = "cat > $dns_reverse_zone_file \<\<EOF\n$dns_reverse_zone\nEOF\n";
+    type_string("$set_dns_reverse_zone");
+
+    my $dns_config_file      = "/etc/named.conf";
+    my $dns_config_file_tmp  = "/etc/named.conf.tmp";
+    my $dns_service_name     = "named";
+    my $get_nameservers_list = "cat $dns_resolver_file | grep -iE \"^nameserver\" | awk \'{print \$NF}\'";
+    my $nameservers_list     = script_output($get_nameservers_list, $wait_script, type_command => 1, proceed_on_failure => 1);
+    my @nameservers_array    = split(/\n+/, $nameservers_list);
+    $nameservers_list = join("; ", @nameservers_array);
+    $nameservers_list = "$nameservers_list" . "; ";
+
+    my $set_dns_forwarders = "sed -i \'s/forwarders\.*/forwarders { $nameservers_list } ;/\' $dns_config_file";
+    assert_script_run("cp $dns_config_file $dns_config_file.orig");
+    assert_script_run("$set_dns_forwarders");
+
+    #The following forward zone info will be inserted into /etc/named.conf
+    #zone "testvirt.net" in {
+    #        type master;
+    #        file "testvirt.net.zone";
+    #};
+    my $set_dns_config_zone = "awk \'done != 1 && /^zone.*\$/ { print \"zone \\\"$dns_domain_forward\\\" in {\\n        type master;\\n        file \\\"$dns_domain_forward\.zone\\\";\\n};\"; done=1 } 1\' $dns_config_file > $dns_config_file_tmp";
+    assert_script_run("$set_dns_config_zone");
+    assert_script_run("mv $dns_config_file_tmp $dns_config_file");
+
+    #The following reverse zone info will be inserted into /etc/named.conf
+    #zone "123.168.192.in-addr.arpa" in {
+    #        type master;
+    #        file "123.168.192.zone";
+    #};
+    $set_dns_config_zone = "awk \'done != 1 && /^zone\.*\$/ { print \"zone \\\"$dns_domain_reverse\.in-addr\.arpa\\\" in {\\n        type master;\\n        file \\\"$dns_domain_reverse\.zone\\\";\\n};\"; done=1 } 1\' $dns_config_file > $dns_config_file_tmp";
+    assert_script_run("$set_dns_config_zone");
+    assert_script_run("mv $dns_config_file_tmp $dns_config_file");
+
+    upload_logs("$dns_resolver_file.orig");
+    upload_logs("$dns_resolver_file");
+    upload_logs("$dns_config_file.orig");
+    upload_logs("$dns_config_file");
+
+    if ($os_installed_release gt '11') {
+        assert_script_run("systemctl enable $dns_service_name");
+        assert_script_run("systemctl restart $dns_service_name");
+        assert_script_run("systemctl status $dns_service_name");
+    }
+    else {
+        assert_script_run("service $dns_service_name restart");
+        assert_script_run("service $dns_service_name status");
+    }
+}
+
+1;
```

Comment:
IMHO, it is better to add the recovery code, which stops the DNS and recovers the settings of dhcp and dns to the backup ones, especially as post_fail_hook, or some calling manually or in other parts of code.